rust/noosphere-storage/src/implementation/rocks_db.rs

```diff
@@ -0,0 +1,158 @@
+#[derive(Clone, Debug)]
+pub struct RocksDbStorage {
+    db: Arc<DbInner>,
+    #[allow(unused)]
+    path: PathBuf,
+}
+
+impl RocksDbStorage {
+    pub fn new(path: PathBuf) -> Result<Self> {
+        std::fs::create_dir_all(&path)?;
+        let canonicalized = path.canonicalize()?;
+        let db = Arc::new(RocksDbStorage::open_db(path)?);
+        Ok(RocksDbStorage {
+            db,
+            path: canonicalized,
+        })
+    }
+
+    async fn get_store(&self, name: &str) -> Result<RocksDbStore> {
+        if SPHERE_DB_STORE_NAMES
+            .iter()
+            .find(|val| **val == name)
+            .is_none()
+        {
+            return Err(anyhow!("No such store named {}", name));
+        }
+
+        RocksDbStore::new(self.db.clone(), name.to_owned())
+    }
+
+    fn open_db(path: PathBuf) -> Result<DbInner> {
+        std::fs::create_dir_all(&path)?;
+        let mut cfs: Vec<ColumnFamilyDescriptor> = Vec::with_capacity(SPHERE_DB_STORE_NAMES.len());
+
+        for store_name in SPHERE_DB_STORE_NAMES {
+            // https://github.com/facebook/rocksdb/wiki/RocksDB-Tuning-Guide
+            let cf_opts = Options::default();
+            cfs.push(ColumnFamilyDescriptor::new(*store_name, cf_opts));
+        }
+
+        let mut db_opts = Options::default();
+        db_opts.create_if_missing(true);
+        db_opts.create_missing_column_families(true);
+
+        Ok(DbInner::open_cf_descriptors(&db_opts, path, cfs)?)
+    }
+}
+
+#[async_trait]
+impl Storage for RocksDbStorage {
+    type BlockStore = RocksDbStore;
+    type KeyValueStore = RocksDbStore;
+
+    async fn get_block_store(&self, name: &str) -> Result<Self::BlockStore> {
+        self.get_store(name).await
+    }
+
+    async fn get_key_value_store(&self, name: &str) -> Result<Self::KeyValueStore> {
+        self.get_store(name).await
+    }
+}
+
+#[derive(Clone)]
+pub struct RocksDbStore {
+    name: String,
+    db: Arc<DbInner>,
+}
+
+impl RocksDbStore {
+    pub fn new(db: Arc<DbInner>, name: String) -> Result<Self> {
+        Ok(RocksDbStore { db, name })
+    }
+
+    /// Returns the column family handle. Unfortunately generated on every call
+    /// due to not being `Sync`, potentially `unsafe` alternatives:
+    /// https://github.com/rust-rocksdb/rust-rocksdb/issues/407
+    fn cf_handle<'a>(&'a self) -> Result<ColumnType> {
+        self.db
+            .cf_handle(&self.name)
+            .ok_or_else(move || anyhow!("Could not open handle for {}", self.name))
+    }
+}
+
+#[async_trait]
+impl Store for RocksDbStore {
+    async fn read(&self, key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let cf = self.cf_handle()?;
+        #[cfg(feature = "rocksdb-mt")]
+        let cf = &cf;
+        Ok(self.db.get_cf(cf, key)?)
+    }
+
+    async fn write(&mut self, key: &[u8], bytes: &[u8]) -> Result<Option<Vec<u8>>> {
+        let cf = self.cf_handle()?;
+        #[cfg(feature = "rocksdb-mt")]
+        let cf = &cf;
+        // @TODO atomicize this
```

Comment:
nit: `// TODO(#issue):`

Reply:
RocksDB only has atomicity for writes, removing this todo